File: src/store/progress.ts
"use client";

import { create } from "zustand";
import { persist, createJSONStorage } from "zustand/middleware";
import { useSyncExternalStore } from "react";
import type { ExamVersion, LicenseId } from "@/lib/types";
import { TT108_DATE } from "@/data/licenses";

export interface QStat {
  /** số lần đúng */
  c: number;
  /** số lần sai */
  w: number;
  /** kết quả lần gần nhất: 1 đúng, 0 sai */
  last: 0 | 1;
  /** thời điểm trả lời gần nhất */
  t: number;
  /** hộp Leitner (0–5) cho ôn tập ngắt quãng */
  box?: number;
  /** thời điểm cần ôn lại (ms) */
  due?: number;
}

/** Khoảng cách ôn lại theo hộp Leitner: 10 phút, 1, 3, 7, 16, 35 ngày. */
const DAY = 86_400_000;
export const REVIEW_INTERVALS = [10 * 60_000, DAY, 3 * DAY, 7 * DAY, 16 * DAY, 35 * DAY];

export interface ExamRecord {
  id: string;
  license: LicenseId;
  /** Số đề trong bộ đề (không có nếu là đề ngẫu nhiên). */
  setNo?: number;
  /** Cấu trúc đề (mặc định tt12). */
  version?: ExamVersion;
  at: number;
  correct: number;
  total: number;
  passed: boolean;
  criticalFail: boolean;
  duration: number;
  wrongIds: number[];
}

interface ProgressState {
  stats: Record<number, QStat>;
  exams: ExamRecord[];
  xp: number;
  streak: { days: number; last: string };
  bestCombo: number;
  bookmarks: number[];
  sound: boolean;
  lastLicense: LicenseId | null;
  /** tên in trên bằng lái ảo */
  driverName: string;
  /** điểm cao nhất thử thách 12 điểm theo hạng */
  arcadeBest: Partial<Record<LicenseId, number>>;
  /** điểm cao nhất mini-game săn biển báo */
  signBest: number;
  /** Cấu trúc đề người dùng chọn (null = tự động theo ngày). */
  examVersion: ExamVersion | null;
  /** Cỡ chữ: 1 (chuẩn), 1.15 (lớn), 1.3 (rất lớn) */
  fontScale: number;
  /** Các bước đã hoàn thành trong lộ trình lấy bằng */
  journey: Record<string, boolean>;
  /** Đã trả lời màn chào (chọn hạng, thời điểm thi) */
  onboarded: boolean;
  /** Dự định thời điểm thi: trước / từ 01/3/2027 / chưa biết */
  examTiming: "before" | "after" | "unknown" | null;
  completeOnboarding: (license: LicenseId, timing: "before" | "after" | "unknown") => void;
  setExamVersion: (v: ExamVersion) => void;
  setFontScale: (n: number) => void;
  toggleJourney: (key: string) => void;
  setDriverName: (n: string) => void;
  setArcadeBest: (id: LicenseId, score: number) => void;
  setSignBest: (score: number) => void;
  record: (qid: number, correct: boolean) => void;
  addXp: (n: number) => void;
  setBestCombo: (n: number) => void;
  addExam: (r: ExamRecord) => void;
  toggleBookmark: (qid: number) => void;
  setSound: (v: boolean) => void;
  setLastLicense: (id: LicenseId) => void;
  reset: () => void;
  importData: (json: string) => boolean;
}

function dayKey(d = new Date()) {
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}-${String(d.getDate()).padStart(2, "0")}`;
}

function nextStreak(s: { days: number; last: string }) {
  const today = dayKey();
  if (s.last === today) return s;
  const y = new Date();
  y.setDate(y.getDate() - 1);
  return { days: s.last === dayKey(y) ? s.days + 1 : 1, last: today };
}

const initial = {
  stats: {} as Record<number, QStat>,
  exams: [] as ExamRecord[],
  xp: 0,
  streak: { days: 0, last: "" },
  bestCombo: 0,
  bookmarks: [] as number[],
  sound: true,
  lastLicense: null as LicenseId | null,
  driverName: "",
  arcadeBest: {} as Partial<Record<LicenseId, number>>,
  signBest: 0,
  examVersion: null as ExamVersion | null,
  fontScale: 1,
  journey: {} as Record<string, boolean>,
  onboarded: false,
  examTiming: null as "before" | "after" | "unknown" | null,
};

export const useProgress = create<ProgressState>()(
  persist(
    (set, get) => ({
      ...initial,
      record: (qid, correct) =>
        set((s) => {
          const prev = s.stats[qid] ?? { c: 0, w: 0, last: 0, t: 0 };
          const now = Date.now();
          const box = correct ? Math.min(REVIEW_INTERVALS.length - 1, (prev.box ?? 0) + (prev.t ? 1 : 2)) : 0;
          return {
            stats: {
              ...s.stats,
              [qid]: {
                c: prev.c + (correct ? 1 : 0),
                w: prev.w + (correct ? 0 : 1),
                last: correct ? 1 : 0,
                t: now,
                box,
                due: now + REVIEW_INTERVALS[box],
              },
            },
            streak: nextStreak(s.streak),
          };
        }),
      addXp: (n) => set((s) => ({ xp: Math.max(0, s.xp + n) })),
      setBestCombo: (n) => set((s) => ({ bestCombo: Math.max(s.bestCombo, n) })),
      addExam: (r) => set((s) => ({ exams: [r, ...s.exams].slice(0, 50), streak: nextStreak(s.streak) })),
      toggleBookmark: (qid) =>
        set((s) => ({
          bookmarks: s.bookmarks.includes(qid) ? s.bookmarks.filter((x) => x !== qid) : [...s.bookmarks, qid],
        })),
      setSound: (v) => set({ sound: v }),
      setDriverName: (n) => set({ driverName: n.slice(0, 32) }),
      setArcadeBest: (id, score) => set((s) => ({ arcadeBest: { ...s.arcadeBest, [id]: Math.max(s.arcadeBest[id] ?? 0, score) } })),
      setSignBest: (score) => set((s) => ({ signBest: Math.max(s.signBest, score) })),
      setExamVersion: (v) => set({ examVersion: v }),
      setFontScale: (n) => set({ fontScale: n }),
      completeOnboarding: (license, timing) =>
        set({
          lastLicense: license,
          examTiming: timing,
          examVersion: timing === "after" ? "tt108" : timing === "before" ? "tt12" : null,
          onboarded: true,
        }),
      toggleJourney: (key) => set((s) => ({ journey: { ...s.journey, [key]: !s.journey[key] } })),
      setLastLicense: (id) => set({ lastLicense: id }),
      reset: () =>
        set({
          ...initial,
          sound: get().sound,
          driverName: get().driverName,
          fontScale: get().fontScale,
          examVersion: get().examVersion,
          onboarded: get().onboarded,
          lastLicense: get().lastLicense,
          examTiming: get().examTiming,
        }),
      importData: (json) => {
        try {
          const data = JSON.parse(json);
          const parsed = sanitizeImport(data?.state ?? data);
          if (!parsed) return false;
          set(parsed);
          return true;
        } catch {
          return false;
        }
      },
    }),
    {
      name: "lai-lua-progress",
      version: 1,
      storage: createJSONStorage(() => localStorage),
      // Giữ dữ liệu cũ khi nâng version (mặc định zustand sẽ bỏ toàn bộ state nếu thiếu migrate).
      migrate: (persisted) => ({ ...initial, ...(persisted as Partial<ProgressState>) }),
      partialize: (s) => ({
        stats: s.stats,
        exams: s.exams,
        xp: s.xp,
        streak: s.streak,
        bestCombo: s.bestCombo,
        bookmarks: s.bookmarks,
        sound: s.sound,
        lastLicense: s.lastLicense,
        driverName: s.driverName,
        arcadeBest: s.arcadeBest,
        signBest: s.signBest,
        examVersion: s.examVersion,
        fontScale: s.fontScale,
        journey: s.journey,
        onboarded: s.onboarded,
        examTiming: s.examTiming,
      }),
    },
  ),
);

const num = (v: unknown, d = 0) => (Number.isFinite(Number(v)) ? Number(v) : d);
const isObj = (v: unknown): v is Record<string, unknown> => typeof v === "object" && v !== null && !Array.isArray(v);

/** Kiểm tra & làm sạch dữ liệu nhập từ tệp JSON — dữ liệu sai kiểu sẽ làm app lỗi ở mọi trang. */
function sanitizeImport(st: unknown): Pick<ProgressState, "stats" | "exams" | "xp" | "streak" | "bestCombo" | "bookmarks"> | null {
  if (!isObj(st) || !isObj(st.stats)) return null;
  const stats: Record<number, QStat> = {};
  for (const [k, v] of Object.entries(st.stats)) {
    const id = Number(k);
    if (!Number.isInteger(id) || !isObj(v)) continue;
    const c = num(v.c);
    const w = num(v.w);
    const box = Math.min(REVIEW_INTERVALS.length - 1, Math.max(0, Math.floor(num(v.box))));
    stats[id] = { c, w, last: num(v.last) ? 1 : 0, t: num(v.t), box, due: v.due === undefined ? undefined : num(v.due) };
  }
  const exams: ExamRecord[] = (Array.isArray(st.exams) ? st.exams : [])
    .filter((e): e is Record<string, unknown> => isObj(e) && typeof e.license === "string")
    .slice(0, 50)
    .map((e) => ({
      id: String(e.id ?? `${e.license}-${num(e.at)}`),
      license: e.license as LicenseId,
      setNo: e.setNo === undefined ? undefined : num(e.setNo),
      version: e.version === "tt108" ? "tt108" : e.version === "tt12" ? "tt12" : undefined,
      at: num(e.at),
      correct: num(e.correct),
      total: num(e.total),
      passed: !!e.passed,
      criticalFail: !!e.criticalFail,
      duration: num(e.duration),
      wrongIds: Array.isArray(e.wrongIds) ? e.wrongIds.map(Number).filter(Number.isInteger) : [],
    }));
  const streak = isObj(st.streak) ? { days: Math.max(0, num(st.streak.days)), last: typeof st.streak.last === "string" ? st.streak.last : "" } : { days: 0, last: "" };
  return {
    stats,
    exams,
    xp: Math.max(0, num(st.xp)),
    streak,
    bestCombo: Math.max(0, num(st.bestCombo)),
    bookmarks: Array.isArray(st.bookmarks) ? [...new Set(st.bookmarks.map(Number).filter(Number.isInteger))] : [],
  };
}

const noop = () => () => {};

/** true sau khi đã chạy ở client (tránh lệch dữ liệu khi hydrate). */
export function useHydrated() {
  return useSyncExternalStore(noop, () => true, () => false);
}

/** Ngày hiện tại có còn giữ chuỗi học hay không. */
export function activeStreak(s: { days: number; last: string }) {
  const today = dayKey();
  const y = new Date();
  y.setDate(y.getDate() - 1);
  return s.last === today || s.last === dayKey(y) ? s.days : 0;
}

/** Cấu trúc đề mặc định theo ngày hiện tại. */
export function defaultExamVersion(now = new Date()): ExamVersion {
  return now >= TT108_DATE ? "tt108" : "tt12";
}

/** Cấu trúc đề đang dùng (người dùng chọn hoặc tự động theo ngày — dùng useNow để server/client khớp khi hydrate). */
export function useExamVersion(): ExamVersion {
  const hydrated = useHydrated();
  const v = useProgress((s) => s.examVersion);
  const now = useNow();
  return (hydrated && v) || defaultExamVersion(new Date(now));
}

let nowMinute = 0;
function readNowMinute() {
  const m = Math.floor(Date.now() / 60_000);
  if (m !== nowMinute) nowMinute = m;
  return nowMinute;
}

/** Thời điểm hiện tại (ms, làm tròn theo phút) — an toàn khi hydrate. */
export function useNow() {
  return useSyncExternalStore(noop, readNowMinute, () => 0) * 60_000;
}

File: ios/Packages/LaiLuaCore/Sources/LaiLuaCore/Progress.swift
import Foundation

/// Thống kê một câu hỏi — cùng tên trường với bản web để nhập/xuất tiến độ qua lại.
public struct QStat: Codable, Hashable, Sendable {
    public var c: Int = 0
    public var w: Int = 0
    /// 1 đúng, 0 sai ở lần gần nhất
    public var last: Int = 0
    public var t: Int64 = 0
    public var box: Int = 0
    public var due: Int64? = nil

    public init(c: Int = 0, w: Int = 0, last: Int = 0, t: Int64 = 0, box: Int = 0, due: Int64? = nil) {
        self.c = c; self.w = w; self.last = last; self.t = t; self.box = box; self.due = due
    }

    private enum K: String, CodingKey { case c, w, last, t, box, due }

    public init(from decoder: Decoder) throws {
        let d = try decoder.container(keyedBy: K.self)
        c = try d.decodeIfPresent(Int.self, forKey: .c) ?? 0
        w = try d.decodeIfPresent(Int.self, forKey: .w) ?? 0
        last = try d.decodeIfPresent(Int.self, forKey: .last) ?? 0
        t = try d.decodeIfPresent(Int64.self, forKey: .t) ?? 0
        box = try d.decodeIfPresent(Int.self, forKey: .box) ?? 0
        due = try d.decodeIfPresent(Int64.self, forKey: .due)
    }
}

public struct ExamRecord: Codable, Hashable, Identifiable, Sendable {
    public let id: String
    public let license: String
    public let setNo: Int?
    public let version: String?
    public let at: Int64
    public let correct: Int
    public let total: Int
    public let passed: Bool
    public let criticalFail: Bool
    public let duration: Int64
    public let wrongIds: [Int]

    public init(id: String, license: String, setNo: Int?, version: String?, at: Int64, correct: Int, total: Int, passed: Bool, criticalFail: Bool, duration: Int64, wrongIds: [Int]) {
        self.id = id; self.license = license; self.setNo = setNo; self.version = version; self.at = at
        self.correct = correct; self.total = total; self.passed = passed; self.criticalFail = criticalFail
        self.duration = duration; self.wrongIds = wrongIds
    }

    private enum K: String, CodingKey { case id, license, setNo, version, at, correct, total, passed, criticalFail, duration, wrongIds }

    public init(from decoder: Decoder) throws {
        let d = try decoder.container(keyedBy: K.self)
        id = try d.decodeIfPresent(String.self, forKey: .id) ?? UUID().uuidString
        license = try d.decode(String.self, forKey: .license)
        setNo = try d.decodeIfPresent(Int.self, forKey: .setNo)
        version = try d.decodeIfPresent(String.self, forKey: .version)
        at = try d.decodeIfPresent(Int64.self, forKey: .at) ?? 0
        correct = try d.decodeIfPresent(Int.self, forKey: .correct) ?? 0
        total = try d.decodeIfPresent(Int.self, forKey: .total) ?? 0
        passed = try d.decodeIfPresent(Bool.self, forKey: .passed) ?? false
        criticalFail = try d.decodeIfPresent(Bool.self, forKey: .criticalFail) ?? false
        duration = try d.decodeIfPresent(Int64.self, forKey: .duration) ?? 0
        wrongIds = try d.decodeIfPresent([Int].self, forKey: .wrongIds) ?? []
    }
}

public struct Streak: Codable, Hashable, Sendable {
    public var days: Int = 0
    public var last: String = ""

    public init(days: Int = 0, last: String = "") { self.days = days; self.last = last }

    private enum K: String, CodingKey { case days, last }

    public init(from decoder: Decoder) throws {
        let d = try decoder.container(keyedBy: K.self)
        days = try d.decodeIfPresent(Int.self, forKey: .days) ?? 0
        last = try d.decodeIfPresent(String.self, forKey: .last) ?? ""
    }
}

/// Toàn bộ tiến độ + cài đặt — cùng định dạng JSON với Android (`progress.json`).
public struct ProgressState: Codable, Hashable, Sendable {
    public var stats: [Int: QStat] = [:]
    public var exams: [ExamRecord] = []
    public var xp: Int = 0
    public var streak = Streak()
    public var bestCombo: Int = 0
    public var bookmarks: [Int] = []
    public var sound: Bool = true
    public var lastLicense: String? = nil
    public var driverName: String = ""
    /// "tt12" | "tt108" | nil (tự động theo ngày)
    public var examVersion: String? = nil
    public var fontScale: Double = 1
    public var onboarded: Bool = false
    /// "before" | "after" | "unknown"
    public var examTiming: String? = nil

    public init() {}

    private enum K: String, CodingKey {
        case stats, exams, xp, streak, bestCombo, bookmarks, sound, lastLicense, driverName, examVersion, fontScale, onboarded, examTiming
    }

    /// JSON lưu `stats` dạng object với khoá là số (giống web) — Codable mặc định của Swift lại
    /// mã hoá `[Int: …]` thành mảng, nên phải tự chuyển khoá.
    public init(from decoder: Decoder) throws {
        let d = try decoder.container(keyedBy: K.self)
        let raw = try d.decodeIfPresent([String: QStat].self, forKey: .stats) ?? [:]
        var stats: [Int: QStat] = [:]
        for (k, v) in raw { if let id = Int(k) { stats[id] = v } }
        self.stats = stats
        exams = try d.decodeIfPresent([ExamRecord].self, forKey: .exams) ?? []
        xp = try d.decodeIfPresent(Int.self, forKey: .xp) ?? 0
        streak = try d.decodeIfPresent(Streak.self, forKey: .streak) ?? Streak()
        bestCombo = try d.decodeIfPresent(Int.self, forKey: .bestCombo) ?? 0
        bookmarks = try d.decodeIfPresent([Int].self, forKey: .bookmarks) ?? []
        sound = try d.decodeIfPresent(Bool.self, forKey: .sound) ?? true
        lastLicense = try d.decodeIfPresent(String.self, forKey: .lastLicense)
        driverName = try d.decodeIfPresent(String.self, forKey: .driverName) ?? ""
        examVersion = try d.decodeIfPresent(String.self, forKey: .examVersion)
        fontScale = try d.decodeIfPresent(Double.self, forKey: .fontScale) ?? 1
        onboarded = try d.decodeIfPresent(Bool.self, forKey: .onboarded) ?? false
        examTiming = try d.decodeIfPresent(String.self, forKey: .examTiming)
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: K.self)
        try c.encode(Dictionary(uniqueKeysWithValues: stats.map { (String($0.key), $0.value) }), forKey: .stats)
        try c.encode(exams, forKey: .exams)
        try c.encode(xp, forKey: .xp)
        try c.encode(streak, forKey: .streak)
        try c.encode(bestCombo, forKey: .bestCombo)
        try c.encode(bookmarks, forKey: .bookmarks)
        try c.encode(sound, forKey: .sound)
        try c.encodeIfPresent(lastLicense, forKey: .lastLicense)
        try c.encode(driverName, forKey: .driverName)
        try c.encodeIfPresent(examVersion, forKey: .examVersion)
        try c.encode(fontScale, forKey: .fontScale)
        try c.encode(onboarded, forKey: .onboarded)
        try c.encodeIfPresent(examTiming, forKey: .examTiming)
    }
}

/// Khoảng cách ôn lại theo hộp Leitner: 10 phút, 1, 3, 7, 16, 35 ngày.
public let REVIEW_INTERVALS: [Int64] = [10 * 60_000, DAY_MS, 3 * DAY_MS, 7 * DAY_MS, 16 * DAY_MS, 35 * DAY_MS]
public let DAY_MS: Int64 = 86_400_000

/// Các phép biến đổi thuần trên `ProgressState` (chuyển từ ProgressStore.kt) — không phụ thuộc nền tảng.
public enum ProgressLogic {
    public static func record(_ s: ProgressState, qid: Int, correct: Bool, now: Int64 = nowMillis(), today: Date = Date()) -> ProgressState {
        var s = s
        let prev = s.stats[qid] ?? QStat()
        let box = correct ? min(REVIEW_INTERVALS.count - 1, prev.box + (prev.t != 0 ? 1 : 2)) : 0
        s.stats[qid] = QStat(
            c: prev.c + (correct ? 1 : 0),
            w: prev.w + (correct ? 0 : 1),
            last: correct ? 1 : 0,
            t: now,
            box: box,
            due: now + REVIEW_INTERVALS[box]
        )
        s.streak = nextStreak(s.streak, today: today)
        return s
    }

    public static func addXp(_ s: ProgressState, _ n: Int) -> ProgressState { var s = s; s.xp = max(0, s.xp + n); return s }
    public static func setBestCombo(_ s: ProgressState, _ n: Int) -> ProgressState { var s = s; s.bestCombo = max(s.bestCombo, n); return s }

    public static func addExam(_ s: ProgressState, _ r: ExamRecord, today: Date = Date()) -> ProgressState {
        var s = s
        s.exams = Array(([r] + s.exams).prefix(50))
        s.streak = nextStreak(s.streak, today: today)
        return s
    }

    public static func toggleBookmark(_ s: ProgressState, _ qid: Int) -> ProgressState {
        var s = s
        if let i = s.bookmarks.firstIndex(of: qid) { s.bookmarks.remove(at: i) } else { s.bookmarks.append(qid) }
        return s
    }

    public static func completeOnboarding(_ s: ProgressState, license: String, timing: String) -> ProgressState {
        var s = s
        s.lastLicense = license
        s.examTiming = timing
        switch timing {
        case "after": s.examVersion = "tt108"
        case "before": s.examVersion = "tt12"
        default: s.examVersion = nil
        }
        s.onboarded = true
        return s
    }

    /// Xoá tiến độ nhưng giữ cài đặt.
    public static func reset(_ s: ProgressState) -> ProgressState {
        var n = ProgressState()
        n.sound = s.sound; n.driverName = s.driverName; n.fontScale = s.fontScale; n.examVersion = s.examVersion
        n.onboarded = s.onboarded; n.lastLicense = s.lastLicense; n.examTiming = s.examTiming
        return n
    }

    // MARK: - Chuỗi ngày học

    public static func dayKey(_ d: Date = Date()) -> String {
        let f = DateFormatter()
        f.calendar = Calendar(identifier: .gregorian)
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = .current
        f.dateFormat = "yyyy-MM-dd"
        return f.string(from: d)
    }

    private static func yesterday(_ d: Date) -> Date { Calendar.current.date(byAdding: .day, value: -1, to: d) ?? d.addingTimeInterval(-86_400) }

    public static func nextStreak(_ s: Streak, today: Date = Date()) -> Streak {
        let t = dayKey(today)
        if s.last == t { return s }
        let y = dayKey(yesterday(today))
        return Streak(days: s.last == y ? s.days + 1 : 1, last: t)
    }

    /// Số ngày học liên tiếp còn hiệu lực.
    public static func activeStreak(_ s: Streak, today: Date = Date()) -> Int {
        let t = dayKey(today)
        let y = dayKey(yesterday(today))
        return (s.last == t || s.last == y) ? s.days : 0
    }

    // MARK: - Xuất / nhập tệp JSON (cùng định dạng với web & Android)

    private struct Export: Encodable {
        let app = "lai-lua"
        let version = 1
        let exportedAt: String
        let state: ProgressState
    }

    /// Chỉ xuất phần tiến độ (không kèm cài đặt) theo đúng định dạng tệp của bản web.
    public static func exportJSON(_ s: ProgressState, at date: Date = Date()) -> String {
        var slim = ProgressState()
        slim.stats = s.stats; slim.exams = s.exams; slim.xp = s.xp; slim.streak = s.streak; slim.bestCombo = s.bestCombo; slim.bookmarks = s.bookmarks
        let enc = JSONEncoder()
        enc.outputFormatting = [.sortedKeys]
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let data = (try? enc.encode(Export(exportedAt: iso.string(from: date), state: slim))) ?? Data()
        return String(decoding: data, as: UTF8.self)
    }

    private struct Envelope: Decodable { let state: ProgressState? }

    /// Nhập tệp JSON (từ app hoặc web) vào `current`. Trả về nil nếu tệp không hợp lệ.
    public static func importJSON(_ text: String, into current: ProgressState) -> ProgressState? {
        let data = Data(text.utf8)
        let dec = JSONDecoder()
        let partial: ProgressState
        if let env = try? dec.decode(Envelope.self, from: data), let st = env.state {
            partial = st
        } else if let st = try? dec.decode(ProgressState.self, from: data) {
            partial = st
        } else {
            return nil
        }
        var s = current
        s.stats = partial.stats
        s.exams = Array(partial.exams.prefix(50))
        s.xp = max(0, partial.xp)
        s.streak = partial.streak
        s.bestCombo = partial.bestCombo
        var seen = Set<Int>()
        s.bookmarks = partial.bookmarks.filter { seen.insert($0).inserted }
        return s
    }
}
